vtex-sites/base.store
# feat: send 'add_to_cart' analytics event

## What's the purpose of this pull request?
This PR uses the methods defined on the store sdk related to sending/receiving analytics events, to send the 'add_to_cart' event. This event is triggered every time a product is added to the shopping cart, and then received by an analytics provider that stores the data. The data that is sent is a merge between the fields required by SAE analytics and by GA4, so it can be used by both.

## How it works? 
The `ProductDetails.ts` and `ProductSummary.ts` passes the required analytics + cart attributes to the `useBuyButton` hook. When the button is pressed, the attributes are sent as an analytics event, which then is listened by the `AnalyticsHandler` and inserted in the `window.dataLayer` array to be consumed by the Analytics service (GA4 or SAE).
After sending the event with the cart attributes plus some extra attributes required by the analytics, the object needs to be destructured and built again so it only has the fields accepted to make the 'add to cart' mutation on the `store-api` side. This is why a new object is build before sending the product to the `addItem` method.

## How to test it?
`yarn test`. New tests added on `cart.test.js`, covering:
- The analytics event is added when clicking on the buy button at the PLP;
- The analytics event is added when clicking on the buy button at the PDP;
- The `add_to_cart` analytics event is sent with at least the minimum required attributes. 
To test it manually, add something into the shopping cart. Then, open the console and check if the `window.dataLayer` array has the correct analytics data for the added item.

## References
https://www.notion.so/vtexhandbook/vtex-addToCart-8125848d75cb4a34adbfa57199fc4b66
https://www.dumky.net/posts/analytics-and-tag-testing-with-cypress/
https://docs.cypress.io/guides/references/assertions#TDD-Assertions
https://docs.cypress.io/guides/references/assertions#BDD-Assertions
https://www.chaijs.com/api/bdd/#method_property

A big thanks to @icazevedo for the help with GA4 and the great documentation of the issues/tasks; also to @tlgimenes for all the help to understand the codebase.
